src/wasm/wasm-type.cpp

```diff
@@ -2456,15 +2467,18 @@ void updateReferencedHeapTypes(
       isTopLevel = false;
       if (type->isRef()) {
         auto ht = type->getHeapType();
+        auto exact = ht.getExactness();
+        ht = ht.with(Inexact);
         if (auto it = canonicalized.find(ht); it != canonicalized.end()) {
-          *type = Type(it->second, type->getNullability());
+          *type = Type(it->second.with(exact), type->getNullability());
```

Comment:
But if exactness is part of the heap type, why do we want this behavior?

Reply:
This is replacing non-canonical heap types with canonical heap types. If `$t'` is the canonical version of `$t`, we need to update uses of `(exact $t)` to be `(exact $t')` instead. The exact type will not appear in the `canonicalized` map because we are canonicalizing type definitions represented by the inexact defined types. `(exact $t)` and `$t` refer to the same type definition, so it would be redundant to store the exact version in the map.

Reply:
I see, thanks. Why does Shared not need to be handled similarly here? Because for non-basic types we store it on HeapTypeInfo?

Reply:
Right. Sharedness is actually part of the type definition, unlike exactness.